src/pipe.test.ts

```diff
@@ -60,6 +60,106 @@ describe('pipe(fn, opts)', () => {
     10 * 1000,
   );
 
+  it(
+    'emits callback in pipeline() after synchronous throw',
+    async () => {
+      const sample = [2, 3, 4, 5, 6];
+      const expected = [2];
+      const actual: number[] = [];
+
+      await new Promise<void>((resolve, reject) => {
+        pipeline(
+          Readable.from(sample),
+          pipe(function (n) {
+            actual.push(n);
+            if (n === 2) {
+              throw new Error(`Fail ${n}`);
+            }
+          }),
+          (err) => {
+            if (err) return resolve();
+            reject(new Error('should not happen'));
+          },
+        );
+      });
+
+      assert.deepStrictEqual(actual, expected);
+    },
+    10 * 1000,
+  );
+
+  it(
+    'emits callback in pipeline() after asynchronous reject',
+    async () => {
+      const sample = [2, 3, 4, 5, 6];
+      const expected = [2];
+      const actual: number[] = [];
+
+      await new Promise<void>((resolve, reject) => {
+        pipeline(
+          Readable.from(sample),
+          pipe(function (n) {
+            return new Promise<void>((resolve, reject) => {
+              setTimeout(() => {
+                actual.push(n);
+                if (n === 2) {
+                  reject(new Error(`Fail ${n}`));
+                }
+                resolve();
+              }, n);
+            });
+          }, 1),
+          (err) => {
+            if (err) return resolve();
+            reject(new Error('should not happen'));
+          },
+        );
+      });
+
+      assert.deepStrictEqual(actual, expected);
+    },
+    10 * 1000,
+  );
+
+  it.skip(
```

Comment:
I'd like to discuss whether we should:
1. wait for all the pending tasks, or
1. stop right after it detects rejection.

The current implementation is 2. So you expect 1., yes? Why?

Reply:
@piglovesyou thanks for your response.

We use pipeline for processing files and commit to disk.
1. Processing files: eg prettier
2. Detect difference between the file in memory and in disk and prompts the user what to do.
3. Commit the file if needed.

- The commit to disk may be pending, so the commit may happen after the pipeline returned and may cause racing conditions.
- The prompt may be waiting for the input, and the pipeline finish and the process continues without releasing the prompt.

The pipeline should complete or fail after every transformation was done to prevent problems.
The benefit of approach 2 is speed, but comes with some drawbacks.

Anyway, this is not as important as the error handling. We wait for the prompt queue to complete at the pipeline end.
Commit racing condition may still happen but should be really rare.

Reply:
Thanks. I see, the data stream arrived in the last step ("commit to disk") should all be handled.

As a nature of the `require('stream').pipeline` spec and Stream itself, I still think the racing condition (2.) is preferable. Let's say we have 3 steps in a pipeline and each of them runs in parallel.

```typescript
pipeline(
    readable,
    pipe(format),
    pipe(prompt),
    pipe(save),
    (err) => {}
)
```

## 1. Pending "save" executions can still aborted even when we wait for them

In a stream pipeline, we don't know when "readable" flows another data. Say we have 3 pending "save" execution and the "format" emits an error.

```typescript
pipeline(
    readable,
    pipe(format),     // 🔥 emits an error
    pipe(prompt),
    pipe(save),        // during 3 "save" pending execution
    (err) => {}
)
```

This case calls the callback without waiting for the 3 "save" executions. A transform can't know the state of another so we can't really wait for the pending steps when we use the Node pipeline.

## 2. We might wait for the pending promises, but there could be data out of the highWatermark waiting to be execution

Say the "save" step has 16 highWatermark and 10 parallel. Now, the "save" has 10 promises to be finished. Again, data can arrive until the buffer gets filled. Even when we wait for the pending promises, there could still be unhandled data, which have been answered as "yes" in "prompt".

```typescript
pipeline(
    readable,
    pipe(format),
    pipe(prompt),   // pushs data to fill the buffer
    pipe(save),        // 🔥  an error occurs during 10 "save" execution. Waiting for all to be finished...
    (err) => {}
)
```

## Suggestion in your case

So how about just ignoring errors in the "save" step? If you want to run all data coming to the "save" step, why don't you do this?

```typescript
pipeline(
    readable,
    pipe(format),
    pipe(prompt),
    pipe(async (data) => save(data).catch(console.error)),
    (err) => {}
)
```

Please correct me if I get it wrong👍

Reply:
@piglovesyou no, I just think there should be a way to know that every processing have been finished.
eg:
```
pipeline(
    readable,
    pipe(format, 5),
    pipe(prompt, 20), // This is the bottleneck
    pipe(save, 5),
    (err) => {}
)
```
If an error is throw, at the moment callback is called, we can have 5 format, 20 prompt, and 5 save operations been executed (pending).
So ideally the callback should be called once every pending operation is settled.

Reply:
I've removed the skipped test.
Doesn't seem too easy to implement.
More discussion is needed, and should be discussed in the issues.
Next step is to know how pipeline behaves in case of exceptions.

thanks @piglovesyou.

Reply:
Oh sorry for my bad. Maybe we should define the terms to discuss, "pending" can have multiple meanings: pending promises per pipe, pending chunks per pipe (meaning number of chunks of highWatermark), and all the incoming chunks per pipeline. I appreciate you take the time for this. A further issue will be welcomed.